botocove/cove_host_account.py

```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
-            .paginate(ChildType="ACCOUNT", ParentId=parent_ou)
-            .build_full_result(),
+
+def _is_organizational_unit_id(resource_id: str) -> bool:
+    return bool(re.match(r"^ou-[0-9a-z]{4,32}-[a-z0-9]{8,32}$", resource_id))
+
+
+def _is_account_id(resource_id: str) -> bool:
+    return bool(re.match(r"^\d{12}$", resource_id))
+
+
+def _list_descendant_accounts(
+    client: OrganizationsClient, ancestor_id: str
+) -> Set[str]:
+    descendant_accounts = _list_accounts_for_parent(client, ancestor_id)
+
+    child_ous = _list_organizational_units_for_parent(client, ancestor_id)
+
+    for ou in child_ous:
+        descendant_accounts.update(_list_descendant_accounts(client, ou))
+
+    return descendant_accounts
+
+
+@lru_cache
+def _list_accounts_for_parent(client: OrganizationsClient, parent_id: str) -> Set[str]:
+    pages = client.get_paginator("list_accounts_for_parent").paginate(
+        ParentId=parent_id
+    )
+
+    accounts = {account["Id"] for page in pages for account in page["Accounts"]}
+
+    return accounts
+
+
+@lru_cache
+def _list_organizational_units_for_parent(
+    client: OrganizationsClient, parent_id: str
+) -> Set[str]:
+    pages = client.get_paginator("list_organizational_units_for_parent").paginate(
+        ParentId=parent_id
+    )
+
+    organizational_units = {
+        organizational_unit["Id"]
+        for page in pages
+        for organizational_unit in page["OrganizationalUnits"]
+    }
+
+    return organizational_units
+
+
+def _map_active_accounts_with_org_metadata(
+    client: OrganizationsClient,
+) -> Dict[str, AccountTypeDef]:
+    pages = client.get_paginator("list_accounts").paginate()
+
+    active_accounts = {
+        account["Id"]: account
+        for page in pages
+        for account in page["Accounts"]
+        if account["Status"] == "ACTIVE"
+    }
+
+    return active_accounts
+
+
+def _map_input_without_org_metadata(
+    account_ids: List[str],
+) -> Dict[str, AccountTypeDef]:
+    """Return a degenerate account map for non-org-accounts. The AccountTypeDef
+    dict has only an Id key."""
+
+    return {_id: AccountTypeDef(Id=_id) for _id in account_ids}
+
+
+def _get_host_account_id(client: STSClient) -> str:
+    host_account_id = client.get_caller_identity()["Account"]
+    return host_account_id
+
+
+def _get_boto3_org_client(
+    assuming_session: Optional[Session],
+    thread_workers: int,
+) -> OrganizationsClient:
+    client: OrganizationsClient = _get_boto3_client(
+        "organizations", assuming_session, thread_workers
+    )
+    return client
+
+
+def _get_boto3_sts_client(
+    assuming_session: Optional[Session],
+    thread_workers: int,
+) -> STSClient:
+    client: STSClient = _get_boto3_client("sts", assuming_session, thread_workers)
+    return client
+
+
+def _get_boto3_client(
+    clientname: Union[Literal["organizations"], Literal["sts"]],
+    assuming_session: Optional[Session],
+    thread_workers: int,
+) -> Any:
+    if assuming_session:
```

Comment:
ah this is the reason for the odd helper function I'd forgotten :) It can almost definitely be done more elegantly but one for another time